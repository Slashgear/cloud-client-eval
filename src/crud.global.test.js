const _ = require('lodash');

const baseUrl = require('../baseUrl');
const request = require('supertest')(baseUrl);

const simpleUser = require('../data/user.json');
const tenUsers = require('../data/tenUsers.json');

describe('Step 1 global REST calls', () => {

    describe('DELETE /user',() => {
        it('should return 200 status', done => {
            request.delete('/user').expect(200, done);
        });

        it('should return empty object', done => {
            request.delete('/user').expect({}, done);
        });
    });

    describe('GET /user', () => {

        beforeEach(done =>
            request.delete('/user').expect(200, () => {
                done();
            })
        );

        it('should return 200 status', done => {
            request.get('/user').expect(200, done);
        });

        it('should return array', done => {
            request.get('/user').expect([],done);
        });

        describe('Given database have 1 users', () => {
            beforeEach(done =>
                request.delete('/user').expect(200, () => {
                    request.post('/user').send(simpleUser).end(done);
                })
            );

            it('should return array with user', done => {
                request.get('/user').end((err,res) => {
                    expect(res.body.length).toEqual(1);
                    expect(_.omit(res.body[0], 'id')).toMatchSnapshot();
                    done();
                });
            });
        });

        describe('Given database have 10 users', () => {
            beforeEach(done =>
                request.delete('/user').expect(200, () => {
                    request.put('/user').send(tenUsers).end(done);
                })
            );

            it('should return array with 10 users', done => {
                request.get('/user').end((err,res) => {
                    expect(res.body.length).toEqual(10);
                    done();
                });
            });

            it('should return users order by', done => {
                request.get('/user').end((err,res) => {
                    const users = _.map(res.body, user => _.omit(user, 'id'));
                    expect(users).toMatchSnapshot();
                    done();
                });
            })
        });
    });

    describe('POST /user', () => {
        beforeEach(done =>
            request.delete('/user').expect(200, () => {
                done();
            })
        );

        it('should return 201 status', done => {
            request.post('/user').send(simpleUser).expect(201, done);
        });

        it('should return object', done => {
            request.post('/user').send(simpleUser).end((err,res) => {
                expect(_.isObject(res.body)).toBe(true);
                done();
            })
        });

        it('should return user with id', done =>{
            request.post('/user').send(simpleUser).end((err,res)=> {
                expect(_.isString(res.body.id)).toBe(true);
                done();
            });
        });


        it('should return same user', done => {
            request.post('/user').send(simpleUser).end((err,res) => {
                expect(_.omit(res.body, 'id')).toMatchSnapshot();
                done();
            })
        });
    });

    describe('PUT /user', () => {
        beforeEach(done =>
            request.delete('/user').expect(200, () => {
                done();
            })
        );

        it('should return 201 status', done => {
            request.put('/user').send(tenUsers).expect(201, done);
        });

        it('should return array', done => {
            request.put('/user').send(tenUsers).end((err,res) => {
                expect(_.isArray(res.body)).toBe(true);
                done();
            });
        });

        it('should return 10 users', done => {
            request.put('/user').send(tenUsers).end((err,res) => {
                expect(res.body.length).toEqual(10);
                done();
            });
        } )
    });
});