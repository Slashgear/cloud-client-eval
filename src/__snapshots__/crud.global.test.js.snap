// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step 1 global REST calls GET /user Given database have 1 users should return array with user 1`] = `
Object {
  "birthDay": "02/24/1994",
  "firstName": "timeo",
  "lastName": "picard",
  "position": Object {
    "lat": -40.962026,
    "lon": 24.684415,
  },
}
`;

exports[`Step 1 global REST calls POST /user should return same user 1`] = `
Object {
  "birthDay": "02/24/1994",
  "firstName": "timeo",
  "lastName": "picard",
  "position": Object {
    "lat": -40.962026,
    "lon": 24.684415,
  },
}
`;
